crates/iota-core/src/authority/shared_object_congestion_tracker.rs

```diff
@@ -137,51 +418,278 @@ impl SharedObjectCongestionTracker {
+        base_types::{ObjectID, SequenceNumber, random_object_ref},
         crypto::{AccountKeyPair, get_key_pair},
+        executable_transaction::VerifiedExecutableTransaction,
         transaction::{CallArg, ObjectArg, VerifiedTransaction},
     };
-    use rstest::rstest;
 
     use super::*;
 
-    fn construct_shared_input_objects(objects: &[(ObjectID, bool)]) -> Vec<SharedInputObject> {
+    // Builds a certificate with a list of shared objects and their mutability. The
+    // certificate is only used to test the SharedObjectCongestionTracker
+    // functions, therefore the content other than shared inputs and gas budget
+    // are not important.
+    pub fn build_transaction(
+        objects: &[(ObjectID, bool)],
+        gas_budget: u64,
+    ) -> VerifiedExecutableTransaction {
+        let (sender, keypair): (_, AccountKeyPair) = get_key_pair();
+        let gas_object = random_object_ref();
+        VerifiedExecutableTransaction::new_system(
+            VerifiedTransaction::new_unchecked(
+                TestTransactionBuilder::new(sender, gas_object, 1000)
+                    .with_gas_budget(gas_budget)
+                    .move_call(
+                        ObjectID::random(),
+                        "unimportant_module",
+                        "unimportant_function",
+                        objects
+                            .iter()
+                            .map(|(id, mutable)| {
+                                CallArg::Object(ObjectArg::SharedObject {
+                                    id: *id,
+                                    initial_shared_version: SequenceNumber::new(),
+                                    mutable: *mutable,
+                                })
+                            })
+                            .collect(),
+                    )
+                    .build_and_sign(&keypair),
+            ),
+            0,
+        )
+    }
+
+    pub(crate) fn initialize_tracker_and_compute_tx_start_time(
+        shared_object_congestion_tracker: &mut SharedObjectCongestionTracker,
+        shared_input_objects: &[SharedInputObject],
+        tx_duration: ExecutionTime,
+    ) -> Option<ExecutionTime> {
+        shared_object_congestion_tracker.initialize_for_shared_objects(shared_input_objects);
+        shared_object_congestion_tracker.compute_tx_start_time(shared_input_objects, tx_duration)
+    }
+
+    pub(crate) fn initialize_tracker_and_try_schedule(
+        shared_object_congestion_tracker: &mut SharedObjectCongestionTracker,
+        cert: &VerifiedExecutableTransaction,
+        max_execution_duration_per_commit: u64,
+        previously_deferred_tx_digests: &HashMap<TransactionDigest, DeferralKey>,
+        commit_round: CommitRound,
+    ) -> SequencingResult {
+        let shared_input_objects: Vec<_> = cert.shared_input_objects().collect();
+        shared_object_congestion_tracker.initialize_for_shared_objects(&shared_input_objects);
```

Comment:
If you change it to this (import should be at the top of the file) you don't need the .collect(), I think there are a few more places like this
```
        use iota_types::transaction::TransactionDataAPI;
        shared_object_congestion_tracker.initialize_for_shared_objects(
            &cert
                .data()
                .inner()
                .intent_message
                .value
                .shared_input_objects(),
        );
```

In the `trait TransactionDataAPI` is `fn shared_input_objects(&self) -> Vec<SharedInputObject>;` this could also return a reference already to not clone the values there, but that's probably out of scope for this PR here

Reply:
Thanks for the suggestion, I replaced all instances with the suggested.